docs/http.md

```diff
@@ -0,0 +1,298 @@
+
+```ruby
+client = Puppet::HTTP::Client.new
+
+session = client.create_session
+session.route_to(:ca) do |route|
+  cert = route.get_certificate('ca')
+  puts "Retrieved cert #{cert.subject.to_utf8} from #{route.server}:#{route.port}"
+end
+```
+
+If an explicit server and port were specified on the command line or
+configuration, such as `puppet agent -t --server foo.example.com`, then the
+`route_to` method will always yield that value.
+
+Otherwise, the session will walk the list of resolvers in priority order:
+
+* DNS SRV
+* Server list
+* Default puppet settings
+
+If the `route_to` method yields a route, but it results in an exception, such as
+"connection refused", then the session will yield the next route, until all
+routes for all resolvers have been tried.
+
+If the caller successfully uses a route, then the session will yield the same
+route the next time `route_to` is called again.
+
+#### DNS SRV
+
+If DNS SRV is enabled, randomly selects one of the targets based on the weight of
+each entry in the SRV record. A target with weight 2 would be twice as likely
+to be chosen as a target with weight 1.
+
+```ruby
+client = Puppet::HTTP::Client.new(use_srv: true, srv_domain: 'puppet.example.com')
+
+session = client.create_session
+session.route_to(:ca) do |route|
+  # route.server is "compiler1.puppet.example.com"
+```
+
+#### Server List
+
+Selects the first available server using puppetserver's simple status endpoint.
+This applies when routing requests to the `:puppet` service, as well as any
+service whose server and port are the same as the `:puppet` service. For
+example, when `:ca_server` and `:report_server` have not been overridden.
+
+```ruby
+client = Puppet::HTTP::Client.new(server_list: [compiler1, compiler2])
+
+session = client.create_session
+session.route_to(:puppet) do |route|
+  # route.server is "compiler1"
+end
+```
+
+#### Default Puppet Settings
+
+Returns a route based on the server and port settings for that service:
+
+| service    | server setting | port setting |
+|------------|----------------|--------------|
+| ca         | ca_server      | ca_port      |
+| fileserver | server         | masterport   |
+| report     | report_server  | report_port  |
+| puppet     | server         | masterport   |
+
+For example, `route_to(:report)` would use `Puppet[:report_server]` and
+`Puppet[:report_port]`.
+
+### Generic HTTP(S) Requests
+
+Puppet agents support downloading file content from 3rd party file servers,
+which reduces load on the compiler. The `Client` will provide a low-level API
+for making `GET` requests for an arbitrary URL, and streaming the response body.
+
+Puppet only trusts the puppet PKI for its REST requests. However, it should be
+possible to additionally trust the system store when making HTTPS requests:
+
+```ruby
+client = Puppet::HTTP::Client.new
+response = http.get_content("https://artifactory.example.com/java.tar.gz", :trust_system_store: true)
```

Comment:
It feels like there will be common modes for trust_system_store: hard true/false on the call, and sometimes depending on a setting.  Not sure if we should get into those details here.

Reply:
For puppet REST calls, we want to only trust the puppet PKI, for example, so we don't accidentally download a malicious catalog. There only case I know for trusting the system store are when downloading file content, or when the ["http" report processor](https://github.com/puppetlabs/puppet/blob/6.7.0/lib/puppet/reports/http.rb#L32). The former is something that will always be enabled. The former will likely be a puppet setting (https://tickets.puppetlabs.com/browse/PUP-7737).